pkg/client/unversioned/request.go

```diff
@@ -427,8 +427,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
 		return r
 	}
 	for k, v := range params {
-		for _, vv := range v {
-			r.setParam(k, vv)
+		for _, value := range v {
+			// TODO: Move it to setParam method, once we get rid of
+			// FieldSelectorParam & LabelSelectorParam methods.
+			if k == unversioned.LabelSelectorQueryParam(r.apiVersion) && value == "" {
+				// Don't set an empty selector for backward compatibility.
```

Comment:
ListOPtions *should* be versioned is my argument.  Seeing this code, seeing
the field transformation code, it's basically versioning.  The API for that
object has to be stable, which is the whole point of what we had "internal"
objects before.  The rest of the path is wire transformation - saying that
there is no wire transformation for ListOptions is silly.  Creating a
different way of handling list options from all the other options is
basically creating an extra code path, not removing a code path.

On Wed, Dec 9, 2015 at 11:13 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/client/unversioned/request.go
> <https://github.com/kubernetes/kubernetes/pull/17836#discussion_r47111463>
> :
>
> > @@ -427,8 +427,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
> >  		return r
> >  	}
> >  	for k, v := range params {
> > -		for _, vv := range v {
> > -			r.setParam(k, vv)
> > +		for _, value := range v {
> > +			// TODO: Move it to setParam method, once we get rid of
> > +			// FieldSelectorParam & LabelSelectorParam methods.
> > +			if k == unversioned.LabelSelectorQueryParam(r.apiVersion) && value == "" {
> > +				// Don't set an empty selector for backward compatibility.
>
> Yes - but we can move it to converter, without getting back to versioned
> ListOptions, I think.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17836/files#r47111463>.
>